import engine
import sql
from PyQt5 import QtCore, QtGui, QtWidgets
from knowledge import major_station
from Ui_zhishiku import Ui_Form


class Ui_MainWindow(object):
    def setupUi(self, MainWindow):
        MainWindow.setObjectName("MainWindow")
        MainWindow.setWindowTitle("TicketAdvisor")  # added
        MainWindow.setWindowIcon(QtGui.QIcon('12306.png'))
        MainWindow.resize(1280, 720)
        self.centralwidget = QtWidgets.QWidget(MainWindow)
        self.centralwidget.setObjectName("centralwidget")
        self.comboBox = QtWidgets.QComboBox(self.centralwidget)
        self.comboBox.setGeometry(QtCore.QRect(470, 80, 91, 22))
        self.comboBox.setObjectName("comboBox")
        self.comboBox.addItem("")
        self.comboBox.addItem("")
        self.comboBox.addItem("")
        self.lineEdit = QtWidgets.QLineEdit(self.centralwidget)
        self.lineEdit.setGeometry(QtCore.QRect(130, 80, 81, 21))
        self.lineEdit.setText("")
        self.lineEdit.setObjectName("lineEdit")
        self.lineEdit_2 = QtWidgets.QLineEdit(self.centralwidget)
        self.lineEdit_2.setGeometry(QtCore.QRect(290, 80, 81, 21))
        self.lineEdit_2.setText("")
        self.lineEdit_2.setObjectName("lineEdit_2")
        self.label = QtWidgets.QLabel(self.centralwidget)
        self.label.setGeometry(QtCore.QRect(220, 85, 15, 15))
        self.label.setObjectName("label")
        self.label_2 = QtWidgets.QLabel(self.centralwidget)
        self.label_2.setGeometry(QtCore.QRect(80, 50, 120, 15))
        self.label_2.setObjectName("label_2")
        self.label_3 = QtWidgets.QLabel(self.centralwidget)
        self.label_3.setGeometry(QtCore.QRect(240, 85, 72, 15))
        self.label_3.setObjectName("label_3")
        self.label_4 = QtWidgets.QLabel(self.centralwidget)
        self.label_4.setGeometry(QtCore.QRect(80, 85, 71, 16))
        self.label_4.setObjectName("label_4")
        self.label_5 = QtWidgets.QLabel(self.centralwidget)
        self.label_5.setGeometry(QtCore.QRect(470, 50, 120, 15))
        self.label_5.setObjectName("label_5")
        self.pushButton = QtWidgets.QPushButton(self.centralwidget)
        self.pushButton.setGeometry(QtCore.QRect(80, 130, 93, 28))
        self.pushButton.setObjectName("pushButton")
        self.pushButton.clicked.connect(self.query)  # added
        self.line_2 = QtWidgets.QFrame(self.centralwidget)
        self.line_2.setGeometry(QtCore.QRect(40, 20, 20, 161))
        self.line_2.setFrameShape(QtWidgets.QFrame.VLine)
        self.line_2.setFrameShadow(QtWidgets.QFrame.Sunken)
        self.line_2.setObjectName("line_2")
        self.line_3 = QtWidgets.QFrame(self.centralwidget)
        self.line_3.setGeometry(QtCore.QRect(1220, 20, 20, 161))
        self.line_3.setFrameShape(QtWidgets.QFrame.VLine)
        self.line_3.setFrameShadow(QtWidgets.QFrame.Sunken)
        self.line_3.setObjectName("line_3")
        self.line_4 = QtWidgets.QFrame(self.centralwidget)
        self.line_4.setGeometry(QtCore.QRect(49, 170, 1181, 20))
        self.line_4.setFrameShape(QtWidgets.QFrame.HLine)
        self.line_4.setFrameShadow(QtWidgets.QFrame.Sunken)
        self.line_4.setObjectName("line_4")
        self.line_5 = QtWidgets.QFrame(self.centralwidget)
        self.line_5.setGeometry(QtCore.QRect(49, 20, 1181, 3))
        self.line_5.setFrameShape(QtWidgets.QFrame.HLine)
        self.line_5.setFrameShadow(QtWidgets.QFrame.Sunken)
        self.line_5.setObjectName("line_5")
        self.label_7 = QtWidgets.QLabel(self.centralwidget)
        self.label_7.setGeometry(QtCore.QRect(240, 220, 81, 16))
        self.label_7.setStyleSheet("color:rgb(71,132,75)")
        self.label_7.setObjectName("label_7")
        self.label_8 = QtWidgets.QLabel(self.centralwidget)
        self.label_8.setGeometry(QtCore.QRect(100, 250, 72, 15))
        self.label_8.setObjectName("label_8")
        self.label_9 = QtWidgets.QLabel(self.centralwidget)
        self.label_9.setGeometry(QtCore.QRect(100, 280, 161, 16))
        self.label_9.setStyleSheet("color:rgb(71,132,75)")
        self.label_9.setObjectName("label_9")
        self.label_10 = QtWidgets.QLabel(self.centralwidget)
        self.label_10.setGeometry(QtCore.QRect(670, 280, 161, 16))
        self.label_10.setStyleSheet("color:rgb(71,132,75)")
        self.label_10.setObjectName("label_10")
        self.label_11 = QtWidgets.QLabel(self.centralwidget)
        self.label_11.setGeometry(QtCore.QRect(100, 220, 131, 16))
        self.label_11.setObjectName("label_11")
        self.line_6 = QtWidgets.QFrame(self.centralwidget)
        self.line_6.setGeometry(QtCore.QRect(50, 200, 1181, 3))
        self.line_6.setFrameShape(QtWidgets.QFrame.HLine)
        self.line_6.setFrameShadow(QtWidgets.QFrame.Sunken)
        self.line_6.setObjectName("line_6")
        self.line_7 = QtWidgets.QFrame(self.centralwidget)
        self.line_7.setGeometry(QtCore.QRect(50, 199, 3, 441))
        self.line_7.setFrameShape(QtWidgets.QFrame.VLine)
        self.line_7.setFrameShadow(QtWidgets.QFrame.Sunken)
        self.line_7.setObjectName("line_7")
        self.line_8 = QtWidgets.QFrame(self.centralwidget)
        self.line_8.setGeometry(QtCore.QRect(1230, 199, 3, 441))
        self.line_8.setFrameShape(QtWidgets.QFrame.VLine)
        self.line_8.setFrameShadow(QtWidgets.QFrame.Sunken)
        self.line_8.setObjectName("line_8")
        self.line_9 = QtWidgets.QFrame(self.centralwidget)
        self.line_9.setGeometry(QtCore.QRect(50, 640, 1181, 3))
        self.line_9.setFrameShape(QtWidgets.QFrame.HLine)
        self.line_9.setFrameShadow(QtWidgets.QFrame.Sunken)
        self.line_9.setObjectName("line_9")
        self.label_6 = QtWidgets.QLabel(self.centralwidget)
        self.label_6.setGeometry(QtCore.QRect(1000, 50, 131, 16))
        self.label_6.setObjectName("label_6")
        self.label_12 = QtWidgets.QLabel(self.centralwidget)
        self.label_12.setGeometry(QtCore.QRect(1000, 85, 45, 15))
        self.label_12.setObjectName("label_12")
        self.lineEdit_3 = QtWidgets.QLineEdit(self.centralwidget)
        self.lineEdit_3.setGeometry(QtCore.QRect(1050, 80, 71, 21))
        self.lineEdit_3.setText("")
        self.lineEdit_3.setObjectName("lineEdit_3")
        self.pushButton_2 = QtWidgets.QPushButton(self.centralwidget)
        self.pushButton_2.setGeometry(QtCore.QRect(1080, 220, 101, 28))
        self.pushButton_2.setObjectName("pushButton_2")
        self.pushButton_2.clicked.connect(self.print_all_res_list)  # added
        self.label_13 = QtWidgets.QLabel(self.centralwidget)
        self.label_13.setGeometry(QtCore.QRect(400, 220, 141, 16))
        self.label_13.setObjectName("label_13")
        self.label_14 = QtWidgets.QLabel(self.centralwidget)
        self.label_14.setGeometry(QtCore.QRect(540, 220, 72, 15))
        self.label_14.setStyleSheet("color:rgb(71,132,75)")
        self.label_14.setObjectName("label_14")
        self.radioButton = QtWidgets.QRadioButton(self.centralwidget)
        self.radioButton.setGeometry(QtCore.QRect(684, 81, 57, 19))
        # self.radioButton.setChecked(True)
        self.radioButton.setObjectName("radioButton")
        self.radioButton_2 = QtWidgets.QRadioButton(self.centralwidget)
        self.radioButton_2.setGeometry(QtCore.QRect(805, 81, 57, 19))
        self.radioButton_2.setObjectName("radioButton_2")
        self.label_17 = QtWidgets.QLabel(self.centralwidget)
        self.label_17.setGeometry(QtCore.QRect(680, 50, 121, 20))
        self.label_17.setObjectName("label_17")
        self.lineEdit_4 = QtWidgets.QLineEdit(self.centralwidget)
        self.lineEdit_4.setGeometry(QtCore.QRect(745, 78, 31, 21))
        self.lineEdit_4.setObjectName("lineEdit_4")
        self.lineEdit_5 = QtWidgets.QLineEdit(self.centralwidget)
        self.lineEdit_5.setGeometry(QtCore.QRect(865, 78, 31, 21))
        self.lineEdit_5.setObjectName("lineEdit_5")
        self.label_15 = QtWidgets.QLabel(self.centralwidget)
        self.label_15.setGeometry(QtCore.QRect(780, 81, 8, 15))
        self.label_15.setObjectName("label_15")
        self.label_16 = QtWidgets.QLabel(self.centralwidget)
        self.label_16.setGeometry(QtCore.QRect(900, 80, 8, 15))
        self.label_16.setObjectName("label_16")
        self.label_18 = QtWidgets.QLabel(self.centralwidget)
        self.label_18.setGeometry(QtCore.QRect(700, 220, 141, 16))
        self.label_18.setObjectName("label_18")
        self.label_19 = QtWidgets.QLabel(self.centralwidget)
        self.label_19.setGeometry(QtCore.QRect(840, 220, 72, 15))
        self.label_19.setObjectName("label_19")
        self.label_19.setStyleSheet("color:rgb(71,132,75)")
        self.tableWidget = QtWidgets.QTableWidget(self.centralwidget)
        self.tableWidget.setGeometry(QtCore.QRect(100, 301, 511, 321))
        self.tableWidget.setObjectName("tableWidget")
        self.tableWidget.setColumnCount(7)  # added
        self.tableWidget.setRowCount(1)
        self.tableWidget.setColumnWidth(0, 105)
        self.tableWidget.setColumnWidth(1, 43)
        self.tableWidget.setColumnWidth(2, 43)
        self.tableWidget.setColumnWidth(3, 82)
        self.tableWidget.setColumnWidth(4, 48)
        self.tableWidget.setColumnWidth(5, 82)
        self.tableWidget.setColumnWidth(6, 47)
        self.tableWidget.setEditTriggers(QtWidgets.QAbstractItemView.NoEditTriggers)
        self.tableWidget_2 = QtWidgets.QTableWidget(self.centralwidget)
        self.tableWidget_2.setGeometry(QtCore.QRect(670, 301, 511, 321))
        self.tableWidget_2.setObjectName("tableWidget_2")
        self.tableWidget_2.setColumnCount(7)  # added
        self.tableWidget_2.setRowCount(1)
        self.tableWidget_2.setColumnWidth(0, 105)
        self.tableWidget_2.setColumnWidth(1, 43)
        self.tableWidget_2.setColumnWidth(2, 43)
        self.tableWidget_2.setColumnWidth(3, 82)
        self.tableWidget_2.setColumnWidth(4, 48)
        self.tableWidget_2.setColumnWidth(5, 82)
        self.tableWidget_2.setColumnWidth(6, 47)
        self.tableWidget_2.setEditTriggers(QtWidgets.QAbstractItemView.NoEditTriggers)
        MainWindow.setCentralWidget(self.centralwidget)
        self.menubar = QtWidgets.QMenuBar(MainWindow)
        self.menubar.setGeometry(QtCore.QRect(0, 0, 1280, 26))
        self.menubar.setObjectName("menubar")
        self.menu_2 = QtWidgets.QMenu(self.menubar)
        self.menu_2.setObjectName("menu_2")
        MainWindow.setMenuBar(self.menubar)
        self.statusbar = QtWidgets.QStatusBar(MainWindow)
        self.statusbar.setObjectName("statusbar")
        MainWindow.setStatusBar(self.statusbar)
        self.actionzhishiku = QtWidgets.QAction(MainWindow)
        self.actionzhishiku.setObjectName("actionzhishiku")
        self.actionzhishiku.triggered.connect(self.printzhishiku)
        self.actionshishiku = QtWidgets.QAction(MainWindow)
        self.actionshishiku.setObjectName("actionshishiku")
        self.menu_2.addAction(self.actionzhishiku)
        self.menu_2.addAction(self.actionshishiku)
        self.menubar.addAction(self.menu_2.menuAction())
        MainWindow.statusBar().addPermanentWidget(QtWidgets.QProgressBar())  # added

        self.retranslateUi(MainWindow)
        QtCore.QMetaObject.connectSlotsByName(MainWindow)

    def retranslateUi(self, MainWindow):
        _translate = QtCore.QCoreApplication.translate
        MainWindow.setWindowTitle(_translate("MainWindow", "TicketAdvisor"))
        self.comboBox.setItemText(0, _translate("MainWindow", "全部"))
        self.comboBox.setItemText(1, _translate("MainWindow", "高铁动车"))
        self.comboBox.setItemText(2, _translate("MainWindow", "普速列车"))
        self.label.setText(_translate("MainWindow", "→"))
        self.label_2.setText(_translate("MainWindow", "您要去什么地方？"))
        self.label_3.setText(_translate("MainWindow", "到达站"))
        self.label_4.setText(_translate("MainWindow", "出发站"))
        self.label_5.setText(_translate("MainWindow", "您喜爱何种座席？"))
        self.pushButton.setText(_translate("MainWindow", "查询"))
        self.label_7.setText(_translate("MainWindow", ""))  # modified
        self.label_8.setText(_translate("MainWindow", "车次一览："))
        self.label_9.setText(_translate("MainWindow", ""))  # modified
        self.label_10.setText(_translate("MainWindow", ""))  # modified
        self.label_11.setText(_translate("MainWindow", "为您推荐的中转地："))
        self.label_6.setText(_translate("MainWindow", "强制中转(可不填)"))
        self.label_12.setText(_translate("MainWindow", "中转地"))
        self.pushButton_2.setText(_translate("MainWindow", "查看所有方案"))
        self.label_13.setText(_translate("MainWindow", "中转总里程(公里)："))
        self.label_14.setText(_translate("MainWindow", ""))  # modified
        self.radioButton.setText(_translate("MainWindow", "里程"))
        self.radioButton_2.setText(_translate("MainWindow", "时间"))
        self.label_17.setToolTip(_translate("MainWindow", "<html><head/><body><p>此处设定如何计算下方窗格中车次的“代价”值。例如，若里程被设定为30%，则代价=30%*里程+70%*时间（绝对数值）。选中单选按钮将把对应项的权重设为100%。</p></body></html>"))
        self.label_17.setText(_translate("MainWindow", "代价的计算方式？"))
        self.label_18.setText(_translate("MainWindow", "中转总时间(分钟)："))
        self.label_19.setText(_translate("MainWindow", ""))  # modified
        self.lineEdit_4.setText(_translate("MainWindow", "100"))
        self.lineEdit_5.setText(_translate("MainWindow", "0"))
        self.label_15.setText(_translate("MainWindow", "%"))
        self.label_16.setText(_translate("MainWindow", "%"))
        self.menu_2.setTitle(_translate("MainWindow", "查看"))
        self.actionzhishiku.setText(_translate("MainWindow", "主要车站列表"))
        self.actionshishiku.setText(_translate("MainWindow", "全国铁路所有车次"))
        self.tableWidget.setHorizontalHeaderLabels(['车次', '里程', '时间', '发站', '发时', '到站', '到时'])  # added
        self.tableWidget_2.setHorizontalHeaderLabels(['车次', '里程', '时间', '发站', '发时', '到站', '到时'])

    def set_table_content(self, table, data):
        table.setRowCount(len(data))
        for i in range(len(data)):
            table.setItem(i, 0, QtWidgets.QTableWidgetItem(str(data[i][0])))
            table.setItem(i, 1, QtWidgets.QTableWidgetItem(str(data[i][2])))
            table.setItem(i, 2, QtWidgets.QTableWidgetItem(str(data[i][3])))
            table.setItem(i, 3, QtWidgets.QTableWidgetItem(str(data[i][4][0])))
            table.setItem(i, 4, QtWidgets.QTableWidgetItem(str(data[i][4][2])))
            table.setItem(i, 5, QtWidgets.QTableWidgetItem(str(data[i][5][0])))
            table.setItem(i, 6, QtWidgets.QTableWidgetItem(str(data[i][5][1])))

    def query(self):
        src = self.lineEdit.text()
        dest = self.lineEdit_2.text()
        w1 = int(self.lineEdit_4.text()) / 100
        qzzz = self.lineEdit_3.text()
        mode = self.comboBox.currentIndex()
        trans_city, cost, dist, time_delta, train1, train2, self.all_list = engine.solve(src, dest, mode, w1, qzzz)
        _translate = QtCore.QCoreApplication.translate
        self.label_7.setText(_translate("MainWindow", trans_city))
        self.label_14.setText(_translate("MainWindow", str(dist)))
        self.label_19.setText(_translate("MainWindow", str(time_delta)))
        self.label_9.setText(_translate("MainWindow", src + '-' + trans_city))
        self.label_10.setText(_translate("MainWindow", trans_city + '-' + dest))
        self.set_table_content(self.tableWidget, train1)
        self.set_table_content(self.tableWidget_2, train2)

    def print_all_res_list(self):
        print(self.all_list)

    def printzhishiku(self):
        # QtWidgets.QMessageBox.about(None, '全国铁路枢纽站列表', str(major_station))
        fm = QtWidgets.QMainWindow()
        f = Ui_Form()
        f.setupUi(fm)
        fm.show()
        # f.setContent(str(major_station))


if __name__ == '__main__':
    sql.init()
    import sys
    app = QtWidgets.QApplication(sys.argv)
    MainWindow = QtWidgets.QMainWindow()
    ui = Ui_MainWindow()
    ui.setupUi(MainWindow)
    MainWindow.show()
    sys.exit(app.exec_())
    sql.close()

